cpp/src/arrow/filesystem/gcsfs_internal.cc

```diff
@@ -62,6 +65,133 @@ Status ToArrowStatus(const google::cloud::Status& s) {
   return Status::OK();
 }
 
+namespace gcs = ::google::cloud::storage;
+
+Result<google::cloud::storage::EncryptionKey> ToEncryptionKey(
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::EncryptionKey{};
+  }
+
+  const auto& keys = metadata->keys();
+  const auto& values = metadata->values();
+
+  for (std::size_t i = 0; i < keys.size(); ++i) {
+    if (keys[i] == "encryptionKeyBase64") {
+      return gcs::EncryptionKey::FromBase64Key(values[i]);
+    }
+  }
+  return gcs::EncryptionKey{};
+}
+
+Result<gcs::KmsKeyName> ToKmsKeyName(
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::KmsKeyName{};
+  }
+
+  const auto& keys = metadata->keys();
+  const auto& values = metadata->values();
+
+  for (std::size_t i = 0; i < keys.size(); ++i) {
+    if (keys[i] == "kmsKeyName") {
+      return gcs::KmsKeyName(values[i]);
+    }
+  }
+  return gcs::KmsKeyName{};
+}
+
+Result<gcs::PredefinedAcl> ToPredefinedAcl(
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::PredefinedAcl{};
+  }
+
+  const auto& keys = metadata->keys();
+  const auto& values = metadata->values();
+
+  for (std::size_t i = 0; i < keys.size(); ++i) {
+    if (keys[i] == "predefinedAcl") {
+      return gcs::PredefinedAcl(values[i]);
+    }
+  }
+  return gcs::PredefinedAcl{};
+}
+
+Result<gcs::WithObjectMetadata> ToObjectMetadata(
+    const std::shared_ptr<const KeyValueMetadata>& metadata) {
+  if (!metadata) {
+    return gcs::WithObjectMetadata{};
+  }
+
+  static auto const setters = [] {
+    using setter = std::function<Status(gcs::ObjectMetadata&, const std::string&)>;
+    return std::map<std::string, setter>{
+        {"cacheControl",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_cache_control(v);
+           return Status::OK();
+         }},
+        {"contentDisposition",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_disposition(v);
+           return Status::OK();
+         }},
+        {"contentEncoding",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_encoding(v);
+           return Status::OK();
+         }},
+        {"contentLanguage",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_language(v);
+           return Status::OK();
+         }},
+        {"contentType",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           m.set_content_type(v);
+           return Status::OK();
+         }},
+        {"customTime",
+         [](gcs::ObjectMetadata& m, const std::string& v) {
+           std::string err;
+           absl::Time t;
+           if (!absl::ParseTime(absl::RFC3339_full, v, &t, &err)) {
```

Comment:
Is absl already an include-time dependency of GCS?

Reply:
Yes.  Some Abseil types are exposed in the public API for `google-cloud-cpp`.